docs/src/developer_docs/model_interface.md

```diff
@@ -0,0 +1,227 @@
+  any `TendencyCallsite` callbacks can modify tendencies before integration.
+
+- `time_step!(model, Δt; callbacks=[])`: advances the model clock and its
+  prognostic variables by one step. Simulation hands in the tuple of
+  `ModelCallsite` callbacks so the model can execute `TendencyCallsite` (before
+  tendencies are applied) and `UpdateStateCallsite` callbacks (after auxiliary
+  updates). The method must call `tick!(model.clock, Δt)` (or equivalent) so
+  that `time(model)` and `iteration(model)` remain consistent.
+
+- `set!(model, kw...)`: not strictly required, but strongly recommended as an
+  interface for users to modify the model's prognostic state.
+
+- `initialize!(model::AbstractModel)`: called exactly once per `run!` before the
+  first time step.
+
+### Optional integrations
+
+While not required for Simulation itself, the following methods enable the rest
+of the Oceananigans ecosystem to “see” the model:
+
+- `fields(model)` and `prognostic_fields(model)`: return `NamedTuple`s of
+  fields so diagnostics, output writers, and NaN checkers know what to touch.
+
+- `default_nan_checker(model)`: customize the `NaNChecker` that Simulation adds
+  by default.
+
+## Models that implement this interface
+
+Several models across the CliMA ecosystem implement this interface:
+
+**Oceananigans.jl**
+- `NonhydrostaticModel`: solves the incompressible Navier-Stokes equations
+- `HydrostaticFreeSurfaceModel`: solves the hydrostatic Boussinesq equations with a free surface
+- `ShallowWaterModel`: solves the shallow water equations
+
+**ClimaOcean.jl**
+- `OceanSeaIceModel`: couples ocean, sea ice, and atmosphere components for Earth system modeling.
+                      The components themselves may be `Simulation` that contain `AbstractModel`,
+                      generating a nested structure.
+
+**ClimaSeaIce.jl**
+- `SeaIceModel`: simulates sea ice thermodynamics and dynamics
+
+**Breeze.jl**
+- `AtmosphereModel`: simulates atmospheric dynamics
+
+## Example: a zero-dimensional `LorenzModel`
+
+The example below shows a deliberately tiny model that integrates the Lorenz
+system on a 0D grid. The implementation demonstrates how little is required:
+store a `grid` and `clock`, provide `time_step!` and `update_state!`
+implementations, and rely on fallbacks for the rest. Here we use an explicit
+forward-Euler step so all of the logic lives directly inside `time_step!`.
+
+### Implementing the interface
+
+```@example model_interface
+using Oceananigans
+using Oceananigans.Models: AbstractModel
+using Oceananigans.Simulations: Simulation, run!
+using Oceananigans.TimeSteppers: Clock, tick!
+import Oceananigans.TimeSteppers: update_state!, time_step!
+import Oceananigans.Fields: set!
+using Oceananigans: TendencyCallsite, UpdateStateCallsite
+
+mutable struct LorenzModel{G, C, P, S} <: AbstractModel{Nothing, Nothing}
+    grid :: G
+    clock :: C
+    parameters :: P
+    state :: S
```

Comment:
this looks a bit Terrarium-like, was that on purpose? :P 

I thought the recommended Oceananigans pattern was to have state variables directly defined as fields on the model struct?